ldap/servers/slapd/daemon.c

```diff
@@ -1116,6 +1164,47 @@ slapd_daemon(daemon_ports_t *ports)
         int select_return = 0;
         PRErrorCode prerr;
 
+#ifdef ENABLE_EPOLL
+        int i;
+        Connection *c;
+        size_t conn_count = g_get_current_conn_count();
+
+        num_poll = epoll_wait(the_connection_table->epollfd, ct_epoll_events, connection_table_size, slapd_wakeup_timer);
+
+        // XXX This makes absolutely no sense, if this following slapi_log_err statement is removed or silenced
```

Comment:
While cleaning up my code to further discussion, I discovered that removing this slapi_log_err or silencing it causes an order of magnitude drop in performance from ~4k/s to ~.6k/s in my testing.  I'm completely befuddled by this, can someone point me in the right direction?

Reply:
C is not a "linear" language. The code here isn't all executing one line at a time, your compiler is free to re-arrange things provided no data dependency conflicts, and your cpu will pipeline and run multiple instructions simultaneously, and in any order it wishes.

One rule in a compiler is about code hoisting/sinking where statements can occur. An if condition is a point in the code where the compiler may NOT re-arrange code and certain statement orders creating ordering. If you remove the slapi_log_err, then you create a scenario where the if condition is dead-code eliminated, meaning that other parts CAN be re-arranged. Without looking at the asm that's emitted and different between the two it's hard to know "why" this is causing such a huge performance drop.

I suspect if you replaced the if condition with:

```
__atomic_thread_fence(__ATOMIC_ACQ_REL)
```

This should cause the "code either side of this fence to remain in those locations". But that also may have it's own issues as well. I've been trying to find a way to prevent code hoist/sink without atomic's but I'm not sure it's possible. 

This is just one of those lovely examples of why C is not a good language in any way. 


Anyway, another option could be instead of the switch statement, to use an if like this:

```
if (num_poll > 0) {
} else if (num_poll < 0 ) {
}
```

That also may end up working nicer than the case switch, especially WRT to optimisation since the "common" case is the first one. You can also then move the if conn_count != last_count inside that loop too.

Anyway, a few ideas for you :)

Reply:
If you suspect it's a thread yield issue, can you try pthread_yield() ?

Reply:
@llile Hey there, I've started to look into this a bit. Looking at my decompiled outputs, I don't see any difference in the results beside the added log output - no weird co-arrangement etc. Even the same variables and offsets. 

Can you provide me:

* What compiler version you are using
* What build settings you are using
* a copy of build/.libs/ns-slapd (should be the ELF file) from the "with" and "without" log message variants so that I can decompile and help investigate?

Reply:
HI @Firstyear, sorry for the delayed response.

The compiler is gcc-11.2.1-7.fc35.x86_64 and the configure settings are   $ ./configure --exec-prefix=/usr --prefix=/ --datarootdir=/usr/share --datadir=/usr/share --libdir=/usr/lib64 --disable-asan --disable-tsan --enable-cmocka --with-systemd --with-selinux --with-openldap

What is the best method to provide you with the binaries?

Have you made any discoveries?

Reply:
I think that google drive or some other drop box would work. Alternately you can tar them into a PR or similar for github. Finally, if they compress to small than 10 MB you can email them to me at william.brown at suse.com :) 

I did discover a bit in the reversing process but nothing obvious that would cause the delay. That's why I'm curious to see your build outputs. I'll also try to build these myself today too and compare once you send me your binaries too.

Reply:
I've had an initial look. My test results with: `/opt/dirsrv/bin/ldclt -h localhost -p 3389 -N 10 -n 4 -e bindeach -e bindonly`.  I avoided search paths so that this would stress only anonymous (fully in memory) binds and connections only. 

```
GCC w_ log line: ldclt[15225]: Global average rate: 9105453.00/thr  (364218.12/sec), total: 36421811
GCC w_out log line: ldclt[22341]: Global average rate: 8829269.00/thr  (353170.75/sec), total: 35317075
Clang w_ log line: ldclt[840]: Global average rate: 9532904.00/thr  (381316.16/sec), total: 38131614
Clang w_out log line: ldclt[27872]: Global average rate: 9507196.00/thr  (380287.84/sec), total: 38028785
```

The two clang results are within `+-0.2%` which means I perceive this to be noise. The GCC result however is `+- 3%`, where the presence of the log line increases performance by 3%. This is still a trivial amount IMO, but there is certainly more weight to the fact that there is a code generation issue with GCC related to the presence or absence of this log line, that does not affect clang. 

I'm not seeing the 20x kind of number that you are reporting though, so I'm wondering if there is something else going on in your test methodology, as my results show it as much closer. Regardless, there is a perceptable difference with the log line with GCC. 

```
GCC w_ log line: ldclt[15225]: Global average rate: 9105453.00/thr  (364218.12/sec), total: 36421811
GCC w_out log line: ldclt[22341]: Global average rate: 8829269.00/thr  (353170.75/sec), total: 35317075
GCC w trace line: Global average rate: 8508577.00/thr  (340343.09/sec), total: 34034307
```

Now if we change the log level to TRACE (so it emits nothing, but still is present) this shows that the performance *decreases*. This means that it is unlikely to be a code placement/alignment issue, but that some behaviour in ldap/servers/slapd/log.c slapi_log_error, after the loglevel check is influencing performance, likely be affecting cache or memory behaviour. Reading that function we can see there are two likely causes. Either the call to acquire the lock, or the call to issue the printf to the console. I attempted removing each of them in seperate builds. 

Removing the either caused the performance to rise to where we expect, which makes me think it is the issuance of the syscall to either printf or lock that is the cause of the performance rise. 

```
GCC w_ log line: ldclt[15225]: Global average rate: 9105453.00/thr  (364218.12/sec), total: 36421811
GCC w_out printf: ldclt[3793]: Global average rate: 9260484.00/thr  (370419.38/sec), total: 37041935
GCC w_out lock: ldclt[10884]: Global average rate: 9313178.00/thr  (372527.12/sec), total: 37252714
```

But if we remove the call to slapd_log_error_proc_internal then the performance increases to the rates we saw without the lock. 

```
GCC w_ log line: ldclt[15225]: Global average rate: 9105453.00/thr  (364218.12/sec), total: 36421811
GCC w_out log_err ldclt[25077]: Global average rate: 9211057.00/thr  (368442.28/sec), total: 36844228
```

This starts to make me think that there is a data dependence or code re-arrangement that's going on by having the call to slapi_log_error present, which impacts the behaviour of the branch predictor potentially. Anyway, so I finally tried putting in the expect and removing the log line:

```
GCC w_ log line: ldclt[15225]: Global average rate: 9105453.00/thr  (364218.12/sec), total: 36421811
GCC w_ log + expect: ldclt[32169]: Global average rate: 9226223.00/thr  (369048.91/sec), total: 36904892
GCC w_out log  + expect: ldclt[6841]: Global average rate: 9100628.00/thr  (364025.12/sec), total: 36402510
```

Adding the builtin expect line as suggested gets us up to the performance with the log line, but the log line *still* is increasing our performance.

Next I tried just simply adding a `printf("");` instead of the log line.

```
GCC w_ log line: ldclt[15225]: Global average rate: 9105453.00/thr  (364218.12/sec), total: 36421811
GCC w_ log + expect: ldclt[32169]: Global average rate: 9226223.00/thr  (369048.91/sec), total: 36904892
GCC w_ printf + expect: ldclt[13949]: Global average rate: 9437884.00/thr  (377515.38/sec), total: 37751534
```

This is starting to get up near clang performance! So we know that it's not the log itself, but likely the act of having the *syscall* that is helping somehow. That's probably also why the interaction with epoll_pr_idle exists.

Trying with sched yield instead of printf/slapi_log_err():

```
GCC w_ log line: ldclt[15225]: Global average rate: 9105453.00/thr  (364218.12/sec), total: 36421811
GCC w_ printf + expect: ldclt[13949]: Global average rate: 9437884.00/thr  (377515.38/sec), total: 37751534
GCC w_ yield + expect: ldclt[28141]: Global average rate: 9817390.00/thr  (392695.59/sec), total: 39269562
```

There we go! Now we are seeing some gains!  The syscall to printf or lock() was likely allowing a yield point for the scheduler, so instead we can manually add one - and it's much cleaner :) 

Now curiously, if we use clang with yield + expect it actually *lowers* performance. Again, likely just related to clang emitting better code. 

So ultimately the changes needed to get GCC to clang's performance are:

```
diff --git a/ldap/servers/slapd/daemon.c b/ldap/servers/slapd/daemon.c
index 6f5d1f6ca..eb9f6f471 100644
--- a/ldap/servers/slapd/daemon.c
+++ b/ldap/servers/slapd/daemon.c
@@ -55,6 +55,7 @@
 #include <private/pprio.h>
 #include <ssl.h>
 #include "fe.h"
+#include <sched.h>

 #if defined(LDAP_IOCP)
 #define SLAPD_WAKEUP_TIMER 250
@@ -1156,12 +1157,22 @@ slapd_daemon(daemon_ports_t *ports)
 #endif

     /* The meat of the operation is in a loop on a call to select */
-    while (!g_get_shutdown()) {
+    // while (!g_get_shutdown()) {
+
+    while (
+#ifdef __clang__
+        !g_get_shutdown()
+#else
+        __builtin_expect(!g_get_shutdown(), !0)
+#endif
+    ) {
 #ifdef ENABLE_EPOLL

         // XXX emitting any message between here and #else, provides 20x performance improvement
         // XXX except as noted below near epoll_pr_idle
-        slapi_log_err(SLAPI_LOG_ERR, "slapd_daemon", "epoll mystery\n");
+#ifndef __clang__
+        sched_yield();
+#endif

         // XXX There appears to be a slight performance improvment with if/else vs. case as firstyear predicted
         if ((num_poll = epoll_wait(the_connection_table->epollfd, ct_epoll_events, connection_table_size, slapd_wakeup_timer)) > 0) {
```